math/fastor/include/algebra/math/impl/fastor_transform3.hpp

```diff
@@ -0,0 +1,234 @@
+
+  /// 4x4 matrix type
+  using matrix44 = Fastor::Tensor<scalar_type, 4, 4>;
+
+  /// Function (object) used for accessing a matrix element
+  using element_getter = algebra::fastor::math::element_getter<scalar_t>;
+
+  /// Size type
+  using size_type = typename matrix_actor_t::size_ty;
+
+  /// Matrix actor
+  using matrix_actor = matrix_actor_t;
+
+  /// 2D Matrix type
+  template <size_type ROWS, size_type COLS>
+  using matrix_type = typename matrix_actor::template matrix_type<ROWS, COLS>;
+
+  /// @}
+
+  /// @name Data objects
+  /// @{
+
+  matrix44 _data;
+  _data.eye2();
+  matrix44 _data_inv;
+  _data_inv.eye2();
+
+  /// @}
+
+  /** Contructor with arguments: t, x, y, z
+   *
+   * @param t the translation (or origin of the new frame)
+   * @param x the x axis of the new frame
+   * @param y the y axis of the new frame
+   * @param z the z axis of the new frame, normal vector for planes
+   *
+   **/
+  ALGEBRA_HOST_DEVICE
+  transform3(const vector3 &t, const vector3 &x, const vector3 &y,
+             const vector3 &z, bool get_inverse = true) {
+
+    _data(Fastor::fseq<0, 3>(), 0) = x;
+    _data(Fastor::fseq<0, 3>(), 1) = y;
+    _data(Fastor::fseq<0, 3>(), 2) = z;
+    _data(Fastor::fseq<0, 3>(), 3) = t;
+
+    if (get_inverse) {
+      _data_inv = Fastor::inverse(_data);
+    }
+  }
+
+  /** Contructor with arguments: t, z, x
+   *
+   * @param t the translation (or origin of the new frame)
+   * @param z the z axis of the new frame, normal vector for planes
+   * @param x the x axis of the new frame
+   *
+   **/
+  ALGEBRA_HOST
+  transform3(const vector3 &t, const vector3 &z, const vector3 &x,
+             bool get_inverse = true)
+      : transform3(t, x, Fastor
+                   : cross(z, x), z, get_inverse) {}
+
+  /** Constructor with arguments: translation
+   *
+   * @param t is the translation
+   **/
+  ALGEBRA_HOST
+  transform3(const vector3 &t) {
+
+    _data(Fastor::fseq<0, 3>(), 3) = t;
+
+    _data_inv = Fastor::inverse(_data);
+  }
+
+  /** Constructor with arguments: matrix
+   *
+   * @param m is the full 4x4 matrix
+   **/
+  ALGEBRA_HOST
+  transform3(const matrix44 &m) {
+    _data = m;
+
+    _data_inv = Fastor::inverse(_data);
+  }
+
+  /** Constructor with arguments: matrix as std::aray of scalar
+   *
+   * @param ma is the full 4x4 matrix asa 16 array
+   **/
+  ALGEBRA_HOST
+  transform3(const array_type<16> &ma) {
+    _data = m;
+
+    _data_inv = Fastor::inverse(_data);
+  }
+
+  /** Default contructors */
+  transform3() = default;
+  transform3(const transform3 &rhs) = default;
+  ~transform3() = default;
+
+  /** Equality operator */
+  ALGEBRA_HOST
+  inline bool operator==(const transform3 &rhs) const {
+
+    return _data == rhs._data;
+  }
+
+  /** This method retrieves the rotation of a transform */
+  ALGEBRA_HOST
+  inline auto rotation() const {
+
+    return Fastor::Tensor<scalar_t, 3, 3>(
+        _data(Fastor::fseq<0, 4>(), Fastor::fseq<0, 4>()));
```

Comment:
The rotation of an affine transformation is the top left 3x3 matrix. Should the sequence not rather go ```Fastor::fseq<0, 3>(), Fastor::fseq<0, 3>()```?

Reply:
Good catch. I've fixed it.